crmd/lrm.c

```diff
@@ -1456,6 +1456,87 @@ synthesize_lrmd_failure(lrm_state_t *lrm_state, xmlNode *action, int rc)
     lrmd_free_event(op);
 }
 
+static void
+record_pending_op(lrm_state_t *lrm_state, xmlNode *action)
+{
+    xmlNode *params = NULL;
+    lrmd_event_data_t *op = NULL;
+    const char *rsc_id = NULL;
+    const char *operation = crm_element_value(action, XML_LRM_ATTR_TASK);
+    const char *target_node = crm_element_value(action, XML_LRM_ATTR_TARGET);
+    xmlNode *xml_rsc = find_xml_node(action, XML_CIB_TAG_RESOURCE, TRUE);
+    lrmd_rsc_info_t rsc;
+
+    if (xml_rsc == NULL) {
+        crm_debug("Skipping recording of pending %s on %s (%p): no resource",
+                 crm_element_value(action, XML_LRM_ATTR_TASK_KEY), target_node, lrm_state);
+        return;
+
+    } else if (operation == NULL) {
+        crm_debug("Skipping recording of pending %s on %s (%p): no operation",
+                 crm_element_value(action, XML_ATTR_TRANSITION_KEY), target_node, lrm_state);
+        return;
+    }
+
+    rsc_id = ID(xml_rsc);
+    CRM_CHECK(rsc_id != NULL, crm_log_xml_err(action, "Bad:action");
+              return);
+
+    if (safe_str_eq(operation, CRMD_ACTION_CANCEL)
+        || safe_str_eq(operation, CRMD_ACTION_DELETE)) {
+        return;
+    }
+
+    params = find_xml_node(action, XML_TAG_ATTRS, TRUE);
+
+    if (params == NULL) {
+        return;
+
+    } else {
+        char *key = crm_meta_name(XML_OP_ATTR_PENDING);
+
+        if (key == NULL) {
+            return;
+
+        } else {
+            const char *record_pending = crm_element_value(params, key);
+
+            free(key);
+
+            if (record_pending == NULL || crm_is_true(record_pending) == FALSE) {
+                return;
+            }
+        }
+    }
+
+    op = construct_op(lrm_state, action, rsc_id, operation);
+    CRM_ASSERT(op != NULL);
+
+    op->call_id = -1;
+    op->op_status = PCMK_LRM_OP_PENDING,
+    op->rc = PCMK_OCF_UNKNOWN;
+
+    op->t_run = time(NULL);
+    op->t_rcchange = op->t_run;
+
+    /* write a "pending" entry to the CIB, inhibit
+     * notification */
+    crm_debug("Recording pending op %s_%s_%d on %s in the CIB", op->rsc_id, op->op_type, op->interval, target_node);
+
+    rsc.id = strdup(op->rsc_id);
+    rsc.type = crm_element_value_copy(xml_rsc, XML_ATTR_TYPE);
+    rsc.class = crm_element_value_copy(xml_rsc, XML_AGENT_ATTR_CLASS);
+    rsc.provider = crm_element_value_copy(xml_rsc, XML_AGENT_ATTR_PROVIDER);
```

Comment:
I don't actually see any side effect of adding the digests in here. ;)  Pengine calculation is not supposed to be invoked when there's still any pending operation anyway.